pkg/action/executor/horizontal_scaler.go

```diff
@@ -0,0 +1,308 @@
+package executor
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+
+	"github.com/vmturbo/kubeturbo/pkg/action/turboaction"
+	"github.com/vmturbo/kubeturbo/pkg/action/util"
+	"github.com/vmturbo/kubeturbo/pkg/discovery/probe"
+	turboscheduler "github.com/vmturbo/kubeturbo/pkg/scheduler"
+	"github.com/vmturbo/kubeturbo/pkg/turbostore"
+
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"github.com/golang/glog"
+)
+
+type HorizontalScaler struct {
+	kubeClient *client.Client
+	broker     turbostore.Broker
+	scheduler  *turboscheduler.TurboScheduler
+}
+
+func NewHorizontalScaler(client *client.Client, broker turbostore.Broker, scheduler *turboscheduler.TurboScheduler) *HorizontalScaler {
+	return &HorizontalScaler{
+		kubeClient: client,
+		broker:     broker,
+		scheduler:  scheduler,
+	}
+}
+
+func (h *HorizontalScaler) Execute(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction, error) {
+	if actionItem == nil {
+		return nil, errors.New("ActionItem passed in is nil")
+	}
+	action, err := h.buildPendingScalingTurboAction(actionItem)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.horizontalScale(action)
+}
+
+func (h *HorizontalScaler) buildPendingScalingTurboAction(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction,
+	error) {
+	targetSE := actionItem.GetTargetSE()
+	targetEntityType := targetSE.GetEntityType()
+	if targetEntityType != proto.EntityDTO_CONTAINER_POD || targetEntityType != proto.EntityDTO_APPLICATION {
```

Comment:
We have similar actionItem validation process for every action type. And every time we check the targetEntityType and actionType. Instead of having this process repeatedly for every action, we should have a method validateAction, which takes two para, allowedEntity and allowedActionType, and in the method, check if this actionItem's targetEntityType and actionType against the two parameters.